hydra-node/src/Hydra/Events/Rotation.hs

```diff
@@ -0,0 +1,104 @@
+
+-- | An EventSource and EventSink combined
+type EventStore e m = (EventSource e m, EventSink e m)
+
+type Checkpointer e = [e] -> e
+
+-- | Creates an event store that rotates according to given config and 'Checkpointer'.
+mkRotatedEventStore ::
+  (HasEventId e, MonadSTM m, MonadUnliftIO m) =>
+  RotationConfig ->
+  Checkpointer e ->
+  LogId ->
+  EventStore e m ->
+  m (EventStore e m)
+mkRotatedEventStore config checkpointer logId eventStore = do
+  logIdV <- newTVarIO logId
+  -- Rules for any event store:
+  -- - sourceEvents will be called in the beginning of the application and whenever the api server wans to load history
+  --   -> might be called multiple times!!
+  -- - putEvent will be called on application start with all events returned by sourceEvents and during processing
+  currentEvents <- getEvents eventSource
+  let currentNumberOfEvents = toInteger $ length currentEvents
+  numberOfEventsV <- newTVarIO currentNumberOfEvents
+  -- XXX: check rotation on startup
+  when (currentNumberOfEvents >= toInteger rotateAfterX) $ do
+    rotateEventLog logIdV numberOfEventsV
+  pure
+    ( EventSource
+        { sourceEvents = rotatedSourceEvents
+        }
+    , EventSink
+        { putEvent = rotatedPutEvent logIdV numberOfEventsV
+        , -- NOTE: Don't allow rotation on-demand
+          rotate = const . const $ pure ()
+        }
+    )
+ where
+  RotateAfter rotateAfterX = config
+  -- TODO: if this turns out to be equal to sourceEvents, then the whole algorithm can just work on each 'EventSink'
+  rotatedSourceEvents = sourceEvents eventSource
+
+  rotatedPutEvent logIdV numberOfEventsV event = do
+    putEvent event
+    -- XXX: bump numberOfEvents
+    numberOfEvents' <- atomically $ do
+      numberOfEvents <- readTVar numberOfEventsV
+      let numberOfEvents' = numberOfEvents + 1
+      writeTVar numberOfEventsV numberOfEvents'
+      pure numberOfEvents'
+    -- XXX: check rotation
+    when (numberOfEvents' >= toInteger rotateAfterX) $ do
+      rotateEventLog logIdV numberOfEventsV
+
+  rotateEventLog logIdV numberOfEventsV = do
+    -- XXX: build checkpoint event
+    history <- getEvents eventSource
+    let checkpoint = checkpointer history
+    -- XXX: rotate with checkpoint
+    currentLogId <- readTVarIO logIdV
+    let currentLogId' = currentLogId + 1
+    rotate currentLogId' checkpoint
+    -- XXX: clear numberOfEvents + bump logId
+    atomically $ do
+      writeTVar numberOfEventsV 0
+      writeTVar logIdV currentLogId'
+
+  (eventSource, EventSink{putEvent, rotate}) = eventStore
+
+prepareRotatedEventStore ::
+  (IsChainState tx, MonadTime m, MonadSTM m, MonadUnliftIO m) =>
+  RotationConfig ->
+  ChainStateType tx ->
+  EventStore (StateEvent tx) m ->
+  m (EventStore (StateEvent tx) m)
+prepareRotatedEventStore rotationConfig initialChainState eventStore = do
+  now <- getCurrentTime
+  let checkpointer = mkChechpointer initialChainState now
+  -- FIXME!
+  let logId = 0
```

Comment:
Yeah.. so where should we start the log file numbering? :)

I see two options that does not require knowing about how many past log files do exist:
- EventId of first or last event (which is strictly increasing)
- Timestamp of when the log was created

Both don't require an initial `LogId` when creating the `EventStore`, but can be found when we decide to do a rotation.

Reply:
Based on our latest discussion, the rotation logic has been refactored to avoid finding the latest logId on startup.

For that reason, we now use the last event id + 1 during rotation, which is also used as the event id for the checkpoint event.

Reply:
Anyhow I'm still tempted to use the timestamp.

wdyt?